pkg/operators/formatters/formatters_test.go

```diff
@@ -0,0 +1,368 @@
+					require.NoError(t, err)
+
+					if datum.annotation != nil {
+						for k, v := range datum.annotation {
+							fa.AddAnnotation(k, v)
+						}
+					}
+
+					lg := logger.DefaultLogger()
+					replacerFunc, err := rpl.replace(lg, ds, fa)
+					require.NoError(t, err)
+
+					data, err := ds.NewPacketSingle()
+					require.NoError(t, err)
+
+					tc.putFunc(fa, data, datum.value)
+
+					err = replacerFunc(data)
+					require.NoError(t, err)
+
+					var formattedField datasource.FieldAccessor
+					if datum.annotation != nil {
+						for _, v := range datum.annotation {
+							formattedField = ds.GetField(v)
+						}
+					} else {
+						formattedField = ds.GetField(tc.name)
+					}
+					require.NotNil(t, formattedField, "formatted field not found for "+tc.name)
+					res, err := tc.getFormattedFunc(formattedField, data)
+					require.NoError(t, err)
+					require.Equal(t, datum.expected, res)
+				},
+			)
+
+		}
+	}
+}
+func TestL3Formatter(t *testing.T) {
+	type testDataL3 struct {
+		rawIP      []byte
+		expectedIP string
+	}
+
+	tests := []testDataL3{
+		{
+			rawIP:      []byte{192, 168, 10, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+			expectedIP: "192.168.10.10",
+		},
+	}
+
+	for _, td := range tests {
+		t.Run(
+			fmt.Sprintf("l3endpoint input: %v expected output: %v", td.rawIP, td.expectedIP),
+			func(t *testing.T) {
+				var rpl replacer
+				found := false
+				for _, r := range replacers {
+					if r.name == "l3endpoint" {
+						rpl = r
+						found = true
+						break
+					}
+				}
+				require.True(t, found, "l3endpoint replacer not found")
+
+				ds, err := datasource.New(datasource.TypeSingle, "l3endpoint")
+				require.NoError(t, err)
+
+				fa, err := ds.AddField("l3endpoint_raw", api.Kind_Bytes)
+				require.NoError(t, err)
+
+				ipField, err := fa.AddSubField("l3endpoint_raw", api.Kind_Bytes, datasource.WithTags("type:"+ebpftypes.IPAddrTypeName))
+				require.NoError(t, err)
+
+				versionField, err := fa.AddSubField("version", api.Kind_Uint8, datasource.WithTags("name:version"))
+				require.NoError(t, err)
+
+				lg := logger.DefaultLogger()
+				replacerFunc, err := rpl.replace(lg, ds, fa)
+				require.NoError(t, err)
+
+				data, err := ds.NewPacketSingle()
+				require.NoError(t, err)
+
+				ipField.PutBytes(data, td.rawIP)
+				versionField.PutUint8(data, 4)
+
+				err = replacerFunc(data)
+				require.NoError(t, err)
+
+				formattedField := ds.GetField("l3endpoint_raw.l3endpoint")
+				require.NotNil(t, formattedField)
+				res, err := formattedField.String(data)
+				require.NoError(t, err)
+				require.Equal(t, td.expectedIP, res)
+			},
+		)
+	}
+}
+
+func TestL4Formatter(t *testing.T) {
+	type testDataL4 struct {
+		rawIP            []byte
+		port             uint16
+		expectedEndpoint string
+		protoNumber      *uint16
+		expectedProto    *string
+	}
+
+	tests := []testDataL4{
+		{
+			rawIP:            []byte{192, 168, 10, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
```

Comment:
Maybe a helper function could be written which intakes a string and gives you the byte slice back.